compiler/util/ir_data_fields.py

```diff
@@ -399,34 +399,37 @@ def __init__(self, oneof: str) -> None:
         super().__init__()
         self.oneof = oneof
         self.owner_type = None
-        self.proxy_name: str = ""
+        self.proxy_name: str = f"_value_{oneof}"
+        self.proxy_choice_name: str = f"which_{oneof}"
         self.name: str = ""
 
     def __set_name__(self, owner, name):
         self.name = name
-        self.proxy_name = f"_{name}"
         self.owner_type = owner
-        # Add our empty proxy field to the class.
+        # Add the empty proxy fields to the class.  This may re-initialize
+        # these if another field in this oneof got there first.
         setattr(owner, self.proxy_name, None)
+        setattr(owner, self.proxy_choice_name, None)
 
     def __get__(self, obj, objtype=None):
-        return getattr(obj, self.proxy_name)
+        if getattr(obj, self.proxy_choice_name, None) == self.name:
+            return getattr(obj, self.proxy_name)
+        else:
+            return None
 
     def __set__(self, obj, value):
         if value is self:
             # This will happen if the dataclass uses the default value, we just
             # default to None.
             value = None
 
-        if value is not None:
-            # Clear the others
```

Comment:
I'm slightly concerned not clearing the others will open us up to serialization going wonky. We should double-check the logic there and add a test where we set the oneof multiple times and confirm the correct oneof is set.

Reply:
The others are implicitly cleared because all of the members in a `oneof` now share the same `which_{oneof_name}` and `_value_{oneof_name}` members.